### DISCUSSION TITLE
Unreal Engine, Anyways of connecting the 3 axis to a PID controller?

### DISCUSSION THREAD

**Discussion by Yaman-dot** [JSBSim-Team/jsbsim#959] [Q&A] [answered]
I'm relatively new to this plugin and I've been exploring it a bit, I was wondering, if is it possible to connect the 3-axis to a PID controller.
Coding the PID controller itself is easy but I don't know a way of connecting it to the Elevators Ailerons and rudder, trying to achieve something like warthunder

**Comment by bcoconni** [comment]
@galleon, @AlbanBERGERET-Epic, @demonixis any hints to provide to @Yaman-dot ?

**Comment by gallonmate** [comment, marked as answer]
Yes anything is possible. More specific details would be useful to answer your question.
Where in Unreal do you want to create the pid controller?
If you want to use Blueprints. Then take a look at the reference project's BP_Airliner. In the main graph there is the Tick event, which then goes to Aileron, Elevator and Rudder Inputs. Those inputs are then being applied in the section right below those nodes. The node "Set members in FlightControlCommands" when then goes to set the Commands struct of the JSBSimMovementComponent. You can connect you pid output in the same way. Everytime the engine ticks, it takes the entire Commmands struct and applys those commands directly to JSBSim.
![image](https://github.com/JSBSim-Team/jsbsim/assets/28702753/b9e21896-ac77-4a76-aef3-0db7734d9064)


If you want to create it in C++, then it's the similar thing, you'd set the appropriate members of the struct variable Command for the JSBSimMovementComponent. Or you wish to re-write the plugin the exact place where the Commands struct gets sent to JSBSim is here:

https://github.com/JSBSim-Team/jsbsim/blob/c9b12c4794046ae1406e1a66941937a64417bfd5/UnrealEngine/Plugins/JSBSimFlightDynamicsModel/Source/JSBSimFlightDynamicsModel/Private/JSBSimMovementComponent.cpp#L592-L604

I'm not sure what feedback values you need or want for you PID. But there is the Aicraft State struct which is current positions of the aircraft control surfaces, copied from JSBSim to UE. Available in c++ and blueprint, it's the same idea as the Commands struct, just in reverse. It's copied from jsbsim. In c++ and blueprints, you can get the struct from the JSBSimMovementComponent. Or if you want to re-write the plugin to get the values directly yourself from JSBSim, it's done here:

https://github.com/JSBSim-Team/jsbsim/blob/c9b12c4794046ae1406e1a66941937a64417bfd5/UnrealEngine/Plugins/JSBSimFlightDynamicsModel/Source/JSBSimFlightDynamicsModel/Private/JSBSimMovementComponent.cpp#L667-L673

Maybe I misunderstood your question. Hopefully that helps a little.

**Reply by Yaman-dot** [reply]
Thank you for helping me out
https://youtu.be/nLM2qWPNqn8
here is a video showcasing my controller with your amazing plugin :)
